### ISSUE TITLE
dnscache goes into busy wait when network connection in unavailable

### ISSUE THREAD

**Issue by deployable** [pjps/ndjbdns#23]
On osx again. If there is no connection to the internet available dnscache pins 1 cpu at 100%

**Comment by pjps** [comment]
Hello Matt,

It looks same as -> https://bugzilla.redhat.com/show_bug.cgi?id=1084747

It seems to occur when the bandwidth is scarce, I haven't been able to reproduce it locally and debug it further. I'm trying with throttling bandwidth to my test machine, but no luck so far. Sorry about the delayed response. Last month was little hectic with travel and festival holidays.

Thank you so much for reporting the issue.

**Comment by deployable** [comment]
That looks like it could be it.

In my case it's lookups that are sent over a specific VPN link. When the
link goes down I get a lot of repeated log entries like:

 + merge 10.x.x.x A whatever.domain.com.
 + merge 10.x.x.x A whatever.domain.com.

Then `dnscache` is pinned to 100%

Chrome seems to have a fairly retry heavy dns lookup algorithm and also
automatically retries failed pages for you, which seems to exacerbate the
problem.

You can probably simulate the same on a linux by using an iptables DROP to
a domains known set of DNS servers and running lookup to that domain.

    iptables -A OUTPUT -p udp -d $ip --dport 53 -j DROP

The maillists seem to hint at a known performance problem with the
MERGEQUERIES algo but this issue occurs a pretty low traffic level. I'll
need to read further.

**Comment by krafts** [comment]
Yes we are facing the same issue across our environment. Rhel 6 servers.

**Comment by timclassic** [comment]
@krafts @deployable Does the latest version of ndjbdns (as of https://github.com/pjps/ndjbdns/commit/73527fbe28c6e18229b28f9d437be0ab5960c21f) resolve your respective issues?
